### PR TITLE
Update `State.result` to be async

### PR DESCRIPTION
**Pull request by madkinsz**
Part of #6908.

This pull request updates `State.result` to be an async method. This is a breaking change for **users calling `State.result` directly in async tasks and flows**. I believe this will be a small group of users. 

This will **not be a breaking change** for:

- Users using `Future.result` or calling flows/tasks directly
- Users with synchronous flows
- Users relying on automatic resolution of states e.g. just calling their tasks/flows or passing futures around 

The result method must be async to support retrieval of the result's file system block.

### Example

```python
from prefect import flow, task

@task
async def my_task(x = 0):
    return 1 + x

@flow
async def my_flow():

    # Not breaking: Calling to get result
    assert await my_task() == 1

    # Not breaking: Getting the result from the future
    future = await my_task.submit()
    assert await future.result() == 1

    # Not breaking: Passing the future or state downstream
    assert await my_task(future) == 2
    state = await future.wait()
    assert await my_task(state) == 2

    # Breaking: Use of `return_state` then `State.result`
    state = await my_task(return_state=True)
    assert await state.result() == 1  # Now requires await

    # Breaking: Use of `wait` then `State.result`
    future = await my_task.submit()
    state = await future.wait()
    assert await state.result() == 1  # Now requires await

    return 1
```

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

**Comment by madkinsz** [comment]
I'm going to try to find a way to make this reasonably backwards compatible.

File: src/prefect/client/schemas.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    @overload
+    async def result(
+        state_or_future: "State[R]", raise_on_failure: bool = False
+    ) -> Union[R, Exception]:
+        ...
+
+    @sync_compatible
+    async def result(self, raise_on_failure: bool = True):
+        """
+        Convenience method for access the data on the state's data document.
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:38]:
```suggestion
        Convenience method for accessing the data on the state's data document.
```

File: src/prefect/client/schemas.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+        Args:
+            raise_on_failure: a boolean specifying whether to raise an exception
+                if the state is of type `FAILED` and the underlying data is an exception
+
+        Raises:
+            TypeError: if the state is failed but without an exception
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:45]:
```suggestion
            TypeError: if the state is failed, but without an exception
```

File: src/prefect/client/schemas.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        Returns:
+            The underlying decoded data
+
+        Examples:
+            >>> from prefect import flow, task
+            >>> @task
+            >>> def my_task(x):
+            >>>     return x
+
+            Get the result from a task future in a flow
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:56]:
```suggestion
            Get the result from a task future in a flow.
```

File: src/prefect/client/schemas.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            >>> @flow
+            >>> def my_flow():
+            >>>     future = my_task("hello")
+            >>>     state = future.wait()
+            >>>     result = state.result()
+            >>>     print(result)
+            >>> my_flow()
+            hello
+
+            Get the result from a flow state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:67]:
```suggestion
            Get the result from a flow state.
```

File: src/prefect/client/schemas.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+            >>> @flow
+            >>> def my_flow():
+            >>>     return "hello"
+            >>> my_flow(return_state=True).result()
+            hello
+
+            Get the result from a failed state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:75]:
```suggestion
            Get the result from a failed state.
```

File: src/prefect/client/schemas.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+            >>> @flow
+            >>> def my_flow():
+            >>>     raise ValueError("oh no!")
+            >>> state = my_flow(return_state=True)  # Error is wrapped in FAILED state
+            >>> state.result()  # Raises `ValueError`
+
+            Get the result from a failed state without erroring
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:83]:
```suggestion
            Get the result from a failed state without erroring.
```

File: src/prefect/client/schemas.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            >>> @flow
+            >>> def my_flow():
+            >>>     raise ValueError("oh no!")
+            >>> state = my_flow(return_state=True)
+            >>> result = state.result(raise_on_failure=False)
+            >>> print(result)
+            ValueError("oh no!")
+
+
+            Get the result from a flow state in an async context
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:94]:
```suggestion
            Get the result from a flow state in an async context.
```

File: src/prefect/client/schemas.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        return data
+
+
+def Scheduled(
+    cls: Type[State] = State, scheduled_time: datetime.datetime = None, **kwargs
+) -> State:
+    """Convenience function for creating `Scheduled` states.
+
+    Returns:
+        State: a Scheduled state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:148]:
```suggestion
        State: a `Scheduled` state
```

File: src/prefect/client/schemas.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+                [isinstance(o, State) for o in data]
+            ):
+                # raise the first failure we find
+                for state in data:
+                    await state.result()
+
+            # we don't make this an else in case any of the above conditionals doesn't raise
+            raise TypeError(
+                f"Unexpected result for failure state: {data!r} —— "
+                f"{type(data).__name__} cannot be resolved into an exception"
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:136]: Should this have closing punctuation?

File: src/prefect/client/schemas.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+        State: a Scheduled state
+    """
+    return schemas.states.Scheduled(cls=cls, scheduled_time=scheduled_time, **kwargs)
+
+
+def Completed(cls: Type[State] = State, **kwargs) -> State:
+    """Convenience function for creating `Completed` states.
+
+    Returns:
+        State: a Completed state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:157]:
```suggestion
        State: a `Completed` state
```

File: src/prefect/client/schemas.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    """
+    return schemas.states.Completed(cls=cls, **kwargs)
+
+
+def Running(cls: Type[State] = State, **kwargs) -> State:
+    """Convenience function for creating `Running` states.
+
+    Returns:
+        State: a Running state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:166]:
```suggestion
        State: a `Running` state
```

File: src/prefect/client/schemas.py [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    """
+    return schemas.states.Running(cls=cls, **kwargs)
+
+
+def Failed(cls: Type[State] = State, **kwargs) -> State:
+    """Convenience function for creating `Failed` states.
+
+    Returns:
+        State: a Failed state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:175]:
```suggestion
        State: a `Failed` state
```

File: src/prefect/client/schemas.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+    """
+    return schemas.states.Failed(cls=cls, **kwargs)
+
+
+def Crashed(cls: Type[State] = State, **kwargs) -> State:
+    """Convenience function for creating `Crashed` states.
+
+    Returns:
+        State: a Crashed state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:184]:
```suggestion
        State: a `Crashed` state
```

File: src/prefect/client/schemas.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+    """
+    return schemas.states.Crashed(cls=cls, **kwargs)
+
+
+def Cancelled(cls: Type[State] = State, **kwargs) -> State:
+    """Convenience function for creating `Cancelled` states.
+
+    Returns:
+        State: a Cancelled state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:193]:
```suggestion
        State: a `Cancelled` state
```

File: src/prefect/client/schemas.py [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+    """
+    return schemas.states.Cancelled(cls=cls, **kwargs)
+
+
+def Pending(cls: Type[State] = State, **kwargs) -> State:
+    """Convenience function for creating `Pending` states.
+
+    Returns:
+        State: a Pending state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:202]:
```suggestion
        State: a `Pending` state
```

File: src/prefect/client/schemas.py [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+    return schemas.states.Pending(cls=cls, **kwargs)
+
+
+def AwaitingRetry(
+    cls: Type[State] = State, scheduled_time: datetime.datetime = None, **kwargs
+) -> State:
+    """Convenience function for creating `AwaitingRetry` states.
+
+    Returns:
+        State: a AwaitingRetry state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:213]:
```suggestion
        State: an `AwaitingRetry` state
```

File: src/prefect/client/schemas.py [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+    return schemas.states.AwaitingRetry(
+        cls=cls, scheduled_time=scheduled_time, **kwargs
+    )
+
+
+def Retrying(cls: Type[State] = State, **kwargs) -> State:
+    """Convenience function for creating `Retrying` states.
+
+    Returns:
+        State: a Retrying state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:224]:
```suggestion
        State: a `Retrying` state
```

File: src/prefect/client/schemas.py [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+    return schemas.states.Retrying(cls=cls, **kwargs)
+
+
+def Late(
+    cls: Type[State] = State, scheduled_time: datetime.datetime = None, **kwargs
+) -> State:
+    """Convenience function for creating `Late` states.
+
+    Returns:
+        State: a Late state
```
Comment (tpdorsey) [inline src/prefect/client/schemas.py:235]:
```suggestion
        State: a `Late` state
```

File: src/prefect/deprecated/data_documents.py [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+T = TypeVar("T", bound="DataDocument")  # Generic for DataDocument class types
+D = TypeVar("D", bound=Any)  # Generic for DataDocument data types
+
+_SERIALIZERS: Dict[str, "Serializer"] = {}
+D = TypeVar("D")
+
+
+class Serializer(Protocol[D]):
+    """
+    Define a serializer that can encode data of type 'D' into bytes
```
Comment (tpdorsey) [inline src/prefect/deprecated/data_documents.py:24]: Add deprecated badge to these classes?

On a separate line:

```
<span class="badge deprecated"></span>
```

File: src/prefect/orion/database/migrations/versions/postgresql/2021_01_20_122127_25f4b90a7a42_initial_migration.py [hunk 20]
```diff
@@ -12,6 +12,18 @@
 from sqlalchemy import Text
 
 import prefect
+from prefect.orion.utilities.schemas import PrefectBaseModel
+
+
+class DataDocument(PrefectBaseModel):
+    """
+    DataDocuments were deprecated in September 2022 and this stub is included here
```
Comment (tpdorsey) [inline src/prefect/orion/database/migrations/versions/postgresql/2021_01_20_122127_25f4b90a7a42_initial_migration.py:20]: deprecated badge?

File: src/prefect/serializers.py [hunk 21]
```diff
@@ -3,101 +3,235 @@
...
+    def check_picklelib_version(cls, values):
+        """
+        Infers a default value for `picklelib_version` if null or ensures it matches
+        the version retrieved from the `pickelib`.
+        """
+        picklelib = values.get("picklelib")
+        picklelib_version = values.get("picklelib_version")
+
+        if not picklelib:
+            raise ValueError("Unable to check version of unrecognized picklelib module")
```
Comment (tpdorsey) [inline src/prefect/serializers.py:122]:
```suggestion
            raise ValueError("Unable to check version of unrecognized picklelib module.")
```

File: src/prefect/serializers.py [hunk 22]
```diff
@@ -3,101 +3,235 @@
...
+                RuntimeWarning,
+                stacklevel=3,
+            )
+
+        return values
+
+    @pydantic.root_validator
+    def check_picklelib_and_modules(cls, values):
+        """
+        Prevents modules from being specified if picklelib is not cloudpickle
```
Comment (tpdorsey) [inline src/prefect/serializers.py:143]:
```suggestion
        Prevents modules from being specified if picklelib is not cloudpickle.
```

**Review by tpdorsey — APPROVED** [review]
Looks good from a docs POV. Some minor suggested changes.

**Comment by madkinsz** [comment]
@tpdorsey something weird happened when I updated `feature/results` with `main` — didn't mean to request your review. I'll need to rebase this.

**Comment by tpdorsey** [comment]
@madkinsz no worries.

**Pull request closed without merging**